source/snapshot/iterator.go

```diff
@@ -0,0 +1,207 @@
+		table:          params.Table,
+		keyColumn:      params.KeyColumn,
+		orderingColumn: params.OrderingColumn,
+		columns:        params.Columns,
+		batchSize:      params.BatchSize,
+	}
+
+	if params.Position != nil {
+		position, err := parsePosition(params.Position)
+		if err != nil {
+			return nil, fmt.Errorf("parse position: %w", err)
+		}
+
+		iterator.position = position
+	}
+
+	iterator.columnTypes, err = coltypes.GetColumnTypes(ctx, iterator.db, iterator.table)
+	if err != nil {
+		return nil, fmt.Errorf("get table column types: %w", err)
+	}
+
+	if err = iterator.loadRows(ctx); err != nil {
+		return nil, fmt.Errorf("load rows: %w", err)
+	}
+
+	return iterator, nil
+}
+
+// HasNext returns a bool indicating whether the iterator has the next record to return or not.
+func (i *Iterator) HasNext(ctx context.Context) (bool, error) {
+	if i.rows.Next() {
+		return true, nil
+	}
+
+	if err := i.loadRows(ctx); err != nil {
+		return false, fmt.Errorf("load rows: %w", err)
+	}
+
+	return false, nil
+}
+
+// Next returns the next record.
+func (i *Iterator) Next(ctx context.Context) (sdk.Record, error) {
+	row := make(map[string]any)
+	if err := i.rows.MapScan(row); err != nil {
+		return sdk.Record{}, fmt.Errorf("scan rows: %w", err)
+	}
+
+	transformedRow, err := coltypes.TransformRow(ctx, row, i.columnTypes)
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("transform row column types: %w", err)
+	}
+
+	i.position = &Position{
+		LastProcessedElementValue: transformedRow[i.orderingColumn],
+	}
+
+	sdkPosition, err := i.position.marshalSDKPosition()
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("marshal position: %w", err)
+	}
+
+	transformedRowBytes, err := json.Marshal(transformedRow)
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("marshal row: %w", err)
+	}
+
+	return sdk.Record{
+		Position:  sdkPosition,
+		CreatedAt: time.Now(),
+		Key: sdk.StructuredData{
+			i.keyColumn: transformedRow[i.keyColumn],
+		},
+		Metadata: map[string]string{
+			metadataKeyTable:  i.table,
+			metadataKeyAction: actionInsert,
+		},
+		Payload: sdk.RawData(transformedRowBytes),
+	}, nil
+}
+
+// Stop closes the underlying db connection.
+func (i *Iterator) Stop(ctx context.Context) error {
+	if i.rows != nil {
+		if err := i.rows.Close(); err != nil {
+			return fmt.Errorf("close rows: %w", err)
+		}
+	}
+
+	if i.db != nil {
+		return i.db.Close()
+	}
+
+	return nil
+}
+
+// loadRows selects a batch of rows from a database, based on the Iterator's
+// table, columns, orderingColumn, batchSize and the current position.
+func (i *Iterator) loadRows(ctx context.Context) error {
+	selectBuilder := sqlbuilder.NewSelectBuilder()
+
+	if len(i.columns) > 0 {
+		selectBuilder.Select(i.columns...)
+	} else {
+		selectBuilder.Select("*")
+	}
+
+	selectBuilder.From(i.table)
+
+	if i.position != nil {
+		selectBuilder.Where(
+			selectBuilder.GreaterThan(i.orderingColumn, i.position.LastProcessedElementValue),
```

Comment:
Is it works for different types ?

Reply:
Yeah, it works with integers, strings, timestamps, etc.